### PR TITLE
[DYOD18/19] Unique Constraints

### PR DESCRIPTION
**Pull request by nilsthamm**
This PR is the final state of the unique constraint groups works. It summarizes the changes from these three PRs: #1416, #1495, #1565

All PRs already have been reviewed, this PR is to enable others an overview of all changes together (@mrzzzrm 😉 )

### CODE REVIEW

File: src/lib/operators/abstract_read_write_operator.hpp [hunk 1]
```diff
@@ -51,6 +51,8 @@ class AbstractReadWriteOperator : public AbstractOperator {
    */
   bool execute_failed() const;
 
+  const std::string table_name();
```
Comment (mrzzzrm) [inline src/lib/operators/abstract_read_write_operator.hpp]: as there is no impl for this function, I'll remove it.

File: src/lib/storage/constraints/base_constraint_checker.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <optional>
+#include <string>
+#include <tuple>
+
+#include "storage/chunk.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+static constexpr std::tuple<bool, ChunkID> CO_TUPLE {true, MAX_CHUNK_ID};
```
Comment (mrzzzrm) [inline src/lib/storage/constraints/base_constraint_checker.hpp]: Never used, I'll remove it

**Comment by nilsthamm** [comment]
@mrzzzrm thanks for the quick fixes!

@mrks From our side, everything is ready. If you don't know anyone else who wants to review our code go ahead and approve/merge it.

**Comment by mrks** [comment]
I believe @mrzzzrm's review is still WIP

**Comment by mrks** [comment]
As discussed in today's round table, I moved this branch to https://github.com/hyrise/hyrise-archive/blob/constraints/master:

In the seminar, we found constraints to be more complicated than initially expected, which is why we spend a lot of time understanding all edge cases. As a result, we have a comprehensive test suite that should cover most problematic cases. Unfortunately, this meant that we lacked time to optimize these constraints. Most importantly, indexes are unused. In the current version, the constraint enforcement destroys all performance that we might have. Also, there seem to be some multi-threading issues remaining.

The constraint definitions from this PR have been pulled as "soft" (i.e., unenforced) constraints into https://github.com/hyrise/hyrise/pull/1854, where they help with optimization.

We still consider this a good start for a future support of hard constraints. Due to a lack of resources, we closed the PR for now, but faithfully hope for the work to be continued at a later point.
